Add the simplified exchange_halo interfaces

pbartholomew08 opened the pull request:
This matches the interface used by u-DALES and allows the user to specify per-axis halo depths when performing an exchange.

The majority of this is based on Sam Owens' work https://egusphere.copernicus.org/preprints/2024/egusphere-2024-96/

This should be merged after #411 and completes the u-DALES integration.

pbartholomew08 committed:
Adding halo exchange interfaces used by uDALES

The halo exchange test is updated to use the full interface in one
axis and the short interface in the other for each pencil orientation

mathrack commented on examples/halo_test/exchange_test.f90:
```diff
@@ -287,10 +287,8 @@ subroutine test_div_haloX()
-                   halo_extents%ys:halo_extents%ye, &
-                   halo_extents%zs:halo_extents%ze))
-      allocate (wh, mold=vh)
+      call alloc_x(vh, opt_global=global, opt_levels=[0, 1, 1])
```
I can extend the alloc interface for a scalar opt_level. If needed ?

pbartholomew08 replied:
I tried something similar recently with gfortran it would always expand the scalar into an array and pad it with invalid values. If you can get it to work then certainly it would be useful

pbartholomew08 replied:
Perhaps we could open this as a new issue if it is desired?

mathrack replied:
See https://github.com/pbartholomew08/2decomp-fft/pull/2

mathrack commented on src/halo.f90:
```diff
@@ -233,13 +238,30 @@ integer function get_pencil(sizes, decomp, opt_pencil)
+
+      levels(:) = level
+      levels(ipencil) = 0
+      print *, "LL ", levels
```
debug code here

pbartholomew08 replied:
Removed

mathrack commented on src/halo.f90:
```diff
@@ -233,13 +238,30 @@ integer function get_pencil(sizes, decomp, opt_pencil)
+
       integer :: s1, s2, s3
 
       s1 = sizes(1)
```
I think sizes must be equal to decomp%xsz / decomp%ysz / decomp%zsz ? The argument could be removed maybe ?

pbartholomew08 replied:
You are right - or at least the behaviour between global/not global might be inconsistent. However this follows the earlier implementation (see equivalent code in `halo_common.f90` in `main` branch)
```
if (ipencil == 1) then  ! X-pencil input
     if (global) then
        xs = decomp%xst(1)
        xe = decomp%xen(1)
        ys = decomp%xst(2) - level
        ye = decomp%xen(2) + level
        zs = decomp%xst(3) - level
        ze = decomp%xen(3) + level
     else
        xs = 1
        xe = s1
        ys = 1 - level
        ye = s2 + level
        zs = 1 - level
        ze = s3 + level
     end if
```
I'd rather not risk introducing new bugs here.

mathrack replied:
This is out of the scope of the present PR. It is discussed in https://github.com/2decomp-fft/2decomp-fft/issues/415. It will be fixed in a dedicated PR.

mathrack commented on src/halo.f90:
```diff
@@ -306,9 +328,54 @@ type(halo_extents_t) function init_halo_extents(ipencil, sizes, decomp, level, g
+        end select
+     end if
+
+     sizes = size(arr) - levels
```
`sizes = size(arr) - 2*levels` ?

pbartholomew08 replied:
Corrected

mathrack commented on src/halo.f90:
```diff
@@ -352,7 +419,43 @@ subroutine halo_exchange_real(arr, ipencil, halo_extents, level, sizes)
+        end select
+     end if
+
+     sizes = size(arr) - levels
```
`sizes = size(arr) - 2*levels` ?

pbartholomew08 replied:
Corrected

pbartholomew08 committed:
Update CHANGELOG

pbartholomew08 committed:
Adding documentation comments to the halo module

pbartholomew08 committed:
Eliminate redundant information in arguments by storing in halo_extents

mathrack committed:
Add opt_depth optional argument in alloc and halo_exchange subroutines (#2)

* Add opt_level optional argument in the alloc subroutines

* Update safety check on opt_level(s) in alloc subroutines

* Add opt_level optional argument in halo_exchange short subroutines

* Split the halo_exchange short interface

* Minor update

* Replace opt_level with opt_depth

* Update exchange_test example

* Fix halo complex

* Fix syntax errors in alloc calls

* Calling halo exchange with namde optional arguments seems to work

In particular gfortan14 would not compile the case when opt_levels
was passed without naming, i.e.

  call halo_exchange(uh, 3, [1, 1, 0])

wouldn't resolve the interface.

I have reduced the test to use the opt_levels and opt_depth calls
for each array - the full interface is exercised by theses calls.

* Eliminate redundant information in arguments by storing in halo_extents

---------

Co-authored-by: Paul Bartholomew <p.bartholomew@epcc.ed.ac.uk>

pbartholomew08 commented on src/halo_exchange_z_body.f90:
```diff
@@ -19,9 +19,10 @@
-  icount = (s2 + 2 * level) * s3
-  ilength = level
-  ijump = s1 + 2 * level
+  ! icount = (s2 + 2 * level(2)) * s3
+  icount = (s2 + 2 * levels(2)) * (s3 + 2 * levels(3))
```
Just to call out this and similar changes for the review. This is ported over from the u-DALES work on 2decomp which allows for more general halos, I need to make sure I understand it myself, however currently it is passing the tests.

mathrack replied:
Is the new interface compatible with 1D halo exchange ? For instance, if the data is in Z pencil, is it possible to perform the halo exchange operation only in Y (case `ipencil = 3` and `levels(1) = 0`) ?

If 1D halo operations are not possible, we could add a safety check when the object halo_extents_t is created.

pbartholomew08 replied:
I'm adding a test of the exchange interface based on slab decomposition

pbartholomew08 replied:
I've begun implementing a test for a slab-based decomposition. It works, however the implementation of the exchanges currently assumes that the depth of the halos is >= 1, I think we should go with the safety check you suggest. We could also add it for non-zero depths in the pencil direction as currently no halo exchange occurs here.

pbartholomew08 replied:
Actually, u-DALES requires non-zero depth in the pencil direction.

pbartholomew08 committed:
Add a halo exchange test for slab decompositions

pbartholomew08 committed:
Check that halo definitions are valid

Currently: the halo depth in pencil direction should be zero and in perpendicular directions >= 1

pbartholomew08 committed:
Allow non-zero halo depths in pencil direction

This may be required, e.g. to fill boundary halo values

pbartholomew08 committed:
Check halo depths aren't negative

pbartholomew08 committed:
Add test case for halo allocation

pbartholomew08 committed:
Minor cleanup of halo exchange test case

pbartholomew08 committed:
Add a constructor for serial test cases

This will make it easier to define unit test for 2decomp library internals

pbartholomew08 committed:
Implement halo buffer size calculation with unit test

pbartholomew08 committed:
Replace hardcoded buffer computation with method calls

This generalises the buffers to arbitrary halo sizes in each axis

pbartholomew08 commented:
I think this is now ready - in addition to any comments/changes requested I would suggest we wait for u-DALES to test this branch before merging

pbartholomew08 committed:
Hardcode halo buffer test expected values rather than duplicate computation

pbartholomew08 committed:
WIP: Add a check that indices are exchanged correctly for halos

pbartholomew08 committed:
Correcting test type for exchange_idx

pbartholomew08 committed:
Test now runs across pencils

pbartholomew08 committed:
Add _GPU preprocessing to exchange_idx test

mathrack committed:
Update alloc_halos.f90

mathrack committed:
Fix alloc_halos.f90

pbartholomew08 committed:
Add missing ACC pragmas

pbartholomew08 committed:
ACC won't let you print device data

We only really need the failing index anyway

pbartholomew08 committed:
ACC kernels weren't defined correctly around if statements

pbartholomew08 committed:
Fix broken ACC pragma

pbartholomew08 committed:
ACC won't print?

pbartholomew08 committed:
Make halo sizes subroutine local

pbartholomew08 committed:
Fix code

pbartholomew08 committed:
Pass grid dimensions to test subroutine

pbartholomew08 committed:
Pass grid dimensions in example

pbartholomew08 committed:
Add ACC data clauses

pbartholomew08 committed:
Add ACC data clause in exchange_idx test

pbartholomew08 committed:
Clean up unused variable in halo_common

pbartholomew08 committed:
Run fprettify
